### PR TITLE
[SPARK-17002][CORE]: Document that spark.ssl.protocol. is required for SSL

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

`spark.ssl.enabled`=true, but failing to set `spark.ssl.protocol` will fail and throw meaningless exception. `spark.ssl.protocol` is required when `spark.ssl.enabled`.

Improvement: require `spark.ssl.protocol` when initializing SSLContext, otherwise throws an exception to indicate that. 

Remove the OrElse("default").

Document this requirement in configure.md
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

Manual tests: 
Build document and check document

Configure `spark.ssl.enabled` only, it throws exception below:
6/08/16 16:04:37 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(mwang); groups with view permissions: Set(); users  with modify permissions: Set(mwang); groups with modify permissions: Set()
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: spark.ssl.protocol is required when enabling SSL connections.
    at scala.Predef$.require(Predef.scala:224)
    at org.apache.spark.SecurityManager.<init>(SecurityManager.scala:285)
    at org.apache.spark.deploy.master.Master$.startRpcEnvAndEndpoint(Master.scala:1026)
    at org.apache.spark.deploy.master.Master$.main(Master.scala:1011)
    at org.apache.spark.deploy.master.Master.main(Master.scala)

Configure `spark.ssl.protocol`  and `spark.ssl.protocol`
It works fine.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 1]
```diff
@@ -282,7 +282,9 @@ private[spark] class SecurityManager(sparkConf: SparkConf)
       }: TrustManager
     })
 
-    val sslContext = SSLContext.getInstance(fileServerSSLOptions.protocol.getOrElse("Default"))
+    require(fileServerSSLOptions.protocol != None,
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: It's fine though you could also write `.isDefined` here, which might be a millimeter more canonical. Some would use `getOrElse` to throw an exception, but `require` seems a little tidier for this purpose. Of course the change's substance looks fine.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 2]
```diff
@@ -282,6 +282,11 @@ private[spark] class SecurityManager(sparkConf: SparkConf)
       }: TrustManager
     })
 
+    if (!fileServerSSLOptions.protocol.isDefined) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: Oh, I liked `require` here, was just suggesting using `isDefined`.
Also the following line now doesn't just call `.get` which was also the right change.

**Comment by wangmiao1981** [comment]
@srowen Thanks for your explanation. I misunderstood your previous comment. Now, I addressed your comments.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 3]
```diff
@@ -282,6 +282,9 @@ private[spark] class SecurityManager(sparkConf: SparkConf)
       }: TrustManager
     })
 
+    require(fileServerSSLOptions.protocol.isDefined,
+      "spark.ssl.protocol is required when enabling SSL connections.")
+
     val sslContext = SSLContext.getInstance(fileServerSSLOptions.protocol.getOrElse("Default"))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: OK, though "getOrElse" is redundant now. "Default" is a meaningless value and I thought that was what the JIRA wanted to also remove.

**Comment by mgummelt** [comment]
I'm fine with this PR, but I'm wondering if it makes more sense to set the default to one of the values here (probably "TLS"): https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#SSLContext

So that the user doesn't have to specify.

**Comment by wangmiao1981** [comment]
@mgummelt I am fine with setting a default value. I have one concern about it. My motivation is to let the user make sure that they choose a proper protocol to use which meets their security requirement. Otherwise, user might think they are using the proper protocol, say TLS1.2, but it actually use TLS. With an exception, we just remind user that you should configure a protocol for secure communication.

**Comment by mgummelt** [comment]
If we document what the default is, I think it's a fair requirement that the user read the docs and verify that it meets their needs.

**Comment by wangmiao1981** [comment]
Jenkins, retest this please.

**Comment by wangmiao1981** [comment]
@srowen Do you have any suggestions on our discussion above? Thanks!

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**